components/membership/membership-service-impl/src/test/kotlin/net/corda/membership/service/impl/CommandsRetryManagerTest.kt

```diff
@@ -0,0 +1,218 @@
+        verify(publisher).start()
+    }
+
+    @Test
+    fun `constructor will start the coordinator`() {
+        verify(coordinator).start()
+    }
+
+    @Test
+    fun `constructor will set the coordinator status to UP`() {
+        verify(coordinator).updateStatus(LifecycleStatus.UP)
+    }
+
+    @Test
+    fun `close will close the publisher`() {
+        manager.close()
+
+        verify(publisher).close()
+    }
+
+    @Test
+    fun `close will close the coordinator`() {
+        manager.close()
+
+        verify(coordinator).close()
+    }
+
+    @Test
+    fun `onPartitionSynced will add a timer for each command`() {
+        val states = (1..3).associate {
+            "key-$it" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            }
+        }
+
+        manager.onPartitionSynced(states)
+
+        verify(coordinator).setTimer(eq("retry-key-1"), eq(2000), any())
+        verify(coordinator).setTimer(eq("retry-key-2"), eq(2000), any())
+        verify(coordinator).setTimer(eq("retry-key-3"), eq(2000), any())
+    }
+
+    @Test
+    fun `onPartitionLost will cancel the timer for each command`() {
+        val states = (1..3).associate {
+            "key-$it" to mock<MembershipAsyncRequestState>()
+        }
+
+        manager.onPartitionLost(states)
+
+        verify(coordinator).cancelTimer("retry-key-1")
+        verify(coordinator).cancelTimer("retry-key-2")
+        verify(coordinator).cancelTimer("retry-key-3")
+    }
+
+    @Test
+    fun `onPostCommit will cancel the timers for removed commands`() {
+        val states = mapOf(
+            "key-1" to null,
+            "key-2" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            },
+            "key-3" to null,
+            "key-4" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            },
+        )
+
+        manager.onPostCommit(states)
+
+        verify(coordinator).cancelTimer("retry-key-1")
+        verify(coordinator).cancelTimer("retry-key-3")
+    }
+
+    @Test
+    fun `onPostCommit will create a timer for new commands commands`() {
+        val states = mapOf(
+            "key-1" to null,
+            "key-2" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            },
+            "key-3" to null,
+            "key-4" to mock<MembershipAsyncRequestState> {
+                on { lastFailedOn } doReturn now
+                on { request } doReturn mock()
+            },
+        )
+
+        manager.onPostCommit(states)
+
+        verify(coordinator).setTimer(eq("retry-key-2"), eq(2000), any())
+        verify(coordinator).setTimer(eq("retry-key-4"), eq(2000), any())
```

Comment:
Similar to previous comment
```suggestion
        verify(coordinator, never()).setTimer(eq("retry-key-1"), eq(2000), any())
        verify(coordinator).setTimer(eq("retry-key-2"), eq(2000), any())
        verify(coordinator, never()).setTimer(eq("retry-key-3"), eq(2000), any())
        verify(coordinator).setTimer(eq("retry-key-4"), eq(2000), any())
```